### PR TITLE
Override `broadcast_object_list` for `torch<1.8`

### LINKED ISSUE #7580: bad device index when broadcasting in multi-node, multi-gpu DDP
## 🐛 Bug

I'm running against Pytorch 1.7.1, and I was getting an "invalid device ordinal" error when running multi-node DDP training. After some digging, it looks like PyTorch was using [rank as the device ID to move a tensor to](https://github.com/pytorch/pytorch/blob/v1.7.1/torch/distributed/distributed_c10d.py#L1371-L1382). This relationship doesn't hold in multi-node, hence the invalid device ordinal. 

After some more digging, I realized Lightning handled this error by [overriding the PyTorch function](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/overrides/torch_distributed.py#L91-L94), but only for versions older than 1.7. Unfortunately, the problem persists in all of 1.7.x. 

If I hack things to always use the override (i.e. check `_TORCH_GREATER_EQUAL_1_8` instead of `_TORCH_GREATER_EQUAL_1_7`), I don't get the error anymore. 

## Please reproduce using the BoringModel

This doesn't have anything to do with a model, but if you want to see some code on how I'm calling PTL, [here you go](https://github.com/exabiome/deep-taxon/blob/summit_doover/bin/bcast_error_mwe.py). 

### To Reproduce

Run [this code](https://github.com/exabiome/deep-taxon/blob/summit_doover/bin/bcast_error_mwe.py) with PyTorch 1.7.x 

### Expected behavior

No "invalid device ordinal" error.

### Environment

 - PyTorch Version (e.g., 1.0): 1.7.1
 - OS (e.g., Linux):  Red Hat Enterprise Linux Server, 7.6 (Maipo)
 - How you installed PyTorch (`conda`, `pip`, source): cloned from IBM's open-ce Conda environment 
 - Build command you used (if compiling from source): N/A
 - Python version: 3.8
 - CUDA/cuDNN version: cudatoolkit-10.2.89, cudnn-7.6.5_10.2
 - GPU models and configuration: V100
 - Any other relevant information: running on PPC

### PR DESCRIPTION
**Pull request by ajtritt**
## What does this PR do?

Fixes #7580

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [n/a] Did you make sure to update the documentation with your changes? (if necessary)
- [ ] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Review by Borda — APPROVED** [review]

File: pytorch_lightning/overrides/torch_distributed.py [hunk 1]
```diff
@@ -88,7 +88,7 @@ def _broadcast_object_list(object_list, src=0, group=None):
             object_list[i] = _tensor_to_object(obj_view, obj_size)
 
 
-if _TORCH_GREATER_EQUAL_1_7 and torch.distributed.is_available():
+if _TORCH_GREATER_EQUAL_1_8 and torch.distributed.is_available():
```
Comment (carmocca) [inline pytorch_lightning/overrides/torch_distributed.py:91]: I don't understand the fix. `_broadcast_object_list` is taken from 1.7.1

https://github.com/PyTorchLightning/pytorch-lightning/blob/20f63377f81f4771d3f128f979b3a0f9b8d219a7/pytorch_lightning/overrides/torch_distributed.py#L43-L44

You say in the linked issue:

> but only for versions older than 1.7. Unfortunately, the problem persists in all of 1.7.x.

`_TORCH_GREATER_EQUAL_1_7` should be true in 1.7.x so we should be using the same implementation as the code was copy-pasted.

Comment (ajtritt) [reply pytorch_lightning/overrides/torch_distributed.py:91]: > I don't understand the fix. _broadcast_object_list is taken from 1.7.1

I think that comment is misleading... It should probably read "Taken from ... and modified"

This is how the original code handles the `to` call:
https://github.com/pytorch/pytorch/blob/1.7/torch/distributed/distributed_c10d.py#L1368-L1369

But this is what the override function does:
https://github.com/PyTorchLightning/pytorch-lightning/blob/20f63377f81f4771d3f128f979b3a0f9b8d219a7/pytorch_lightning/overrides/torch_distributed.py#L59-L65

> _TORCH_GREATER_EQUAL_1_7 should be true in 1.7.x 
Right. This is the problem. The override should not happen if Torch is 1.7.x. If Torch is above 1.7.x (i.e. 1.8.0 and up), you can use the Torch function.

**Review by carmocca — APPROVED** [review]

**Review by tchaton — APPROVED** [review]
LGTM !

**Review by awaelchli — APPROVED** [review]
